.github/workflows/build_and_test.yaml

```diff
@@ -0,0 +1,164 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+name: Build and Run Tests
+
+on:
+  push:
+    # Publish only on `master`
+    branches:
+      - master
+  pull_request:
+    branches:
+      - master
+  release:
+    types: [published, edited]
+
+jobs:
+  build:
+    name: Build repository
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out the repo
+        uses: actions/checkout@v2
+      - name: Set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      # Stores external dependencies, can be further improved with Gradle 6.1
+      - name: Cache Gradle Dependencies
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.gradle/caches
+            ~/.gradle/wrapper
+          # Only rebuild cache if build.gradle is changed
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle') }}
+          restore-keys: ${{ runner.os }}-gradle
+      - name: Build repository
+        run: |
+          ./gradlew --no-daemon clean build -x test -x javadoc -x findbugsMain -x findbugsTest -x checkstyleMain -x checkstyleJmh -x checkstyleTest -Dorg.gradle.parallel=true
+
+  test_coverage:
+    runs-on: ubuntu-latest
+    name: Generate test coverage
+    needs: build
+    steps:
```

Comment:
What is the benefit of having multiple jobs, compared to having multiple steps (build/test/coverage) in a single job?
I see that those jobs have a lot of repeated configs, and people might update it in one place, and forget to update it in other jobs.

Reply:
Jobs have the ability to be run in parallel, and steps do not. Since the test coverage and static checks are independent of the unit tests, it would be better to separate them from the normal build process to speed up the pipeline.

I agree that the configs are repeated a lot. It is possible to write composite steps in Github actions so that we can reuse them in multiple jobs though. https://docs.github.com/en/actions/creating-actions/creating-a-composite-run-steps-action. I can implement it in this PR or put it in a different PR, depending on how high you want to prioritize it.

Reply:
Responding to your first comment,

Yup that's a good point, I made the other jobs dependent on the first build job so now it should only download dependencies once. The dependencies should be stored by the gradle cache after `build` which is shared across jobs, but if it's wildcard then gradle might force a dependency download/check. I think in this scenario if it causes a problem it's acceptable as the later steps would have the newer dependency, and if it causes the tests to break then we should really be building against an exact version.

Reply:
ok, we can have separate jobs for now, and later you can refactor them to use composite steps that hold common parts.